doc/man/man3/sockets.3.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man3/sockets.3.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH SOCKETS 3 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Library Calls"
.\"
.\"
.SH NAME
.B libsocket, sockets
\- Linux Fast-STREAMS Sockets Library
.\"
.\"
.SH SYNOPSIS
.PP
.ad l
.B #include <sys/socket.h>
.HP 12
.BI "int accept(int " socket ", struct sockaddr *" address ", socklen_t *" address_len );
.PD 0
.HP
.BI "int bind(int " socket ", const struct sockaddr *" address ", socklen_t " address_len );
.HP
.BI "int connect(int " socket ", const struct sockaddr *" address ", socklen_t " address_len );
.HP
.BI "int getpeername(int " socket ", struct sockaddr *" address ", socklen_t *" address_len );
.HP
.BI "int getsockname(int " socket ", struct sockaddr *" address ", socklen_t *" address_len );
.HP
.BI "int getsockopt(int " socket ", int " level ", int " option_name ", void *" option_value ", socklen_t *" option_len );
.HP
.BI "int listen(int " socket ", int " backlog );
.HP
.BI "ssize_t recv(int " socket ", void *" buffer ", size_t " length ", int " flags );
.HP
.BI "ssize_t recvfrom(int " socket ", void *" buffer ", size_t " length ", int " flags ", struct sockaddr *" address ", socklen_t *" address_len );
.HP
.BI "ssize_t recvmsg(int " socket ", struct msghdr *" message ", int " flags );
.HP
.BI "ssize_t send(int " socket ", const void *" buffer ", size_t " length ", int " flags );
.HP
.BI "ssize_t sendmsg(int " socket ", const struct msghdr *" message ", int " flags );
.HP
.BI "ssize_t sendto(int " socket ", const void *" message ", size_t " length ", int " flags ", const struct sockaddr *" dest_addr ", socklen_t " dest_len );
.HP
.BI "int setsockopt(int " socket ", int " level ", int " option_name ", const void *" option_value ", socklen_t " option_len );
.HP
.BI "int shutdown(int " socket ", int " how );
.HP
.BI "int socket(int " domain ", int " type ", int " protocol );
.HP
.BI "int socketpair(int " domain ", int " type ", int " protocol ", int " socket_vector[2] );
.PD
.ad b
.\"
.\"
.SH OVERVIEW
.PP
.B libsocket
is a sockets compatibility library for
.BR "Linux Fast-STREAMS" .
It works in conjunction with a cooperating
.BR socksys (4)
driver
and
.BR sockmod (4)
module.
The
.BR socksys (4)
driver provides the ability to open Streams conforming to the
.IR "Transport Provider Interface" ,
.BR tpi (7).
The
.BR sockmod (4)
module provides the ability to transform TPI Streams into sockets.
.\"
.\"
.SH DESCRIPTION
.PP
.B libsocket
is a Linux
.IR STREAMS (4)
socket library that uses the
.BR sockmod (4)
module or
.BR socksys (4)
driver.
It implements the following sockets API library functions:
.TP \w'\fBgetsockname\fR(3)\ 'u
.BR accept (3)
accept a new connection on a socket
.PD 0
.TP
.BR bind (3)
bind a name to a socket.
.TP
.BR connect (3)
connnect a socket.
.TP
.BR getpeername (3)
get the name of the peer socket.
.TP
.BR getsockname (3)
get the socket name.
.TP
.BR getsockopt (3)
get the socket options.
.TP
.BR listen (3)
listen for socket connections and limit the queue of incoming connections.
.TP
.BR recv (3)
receive a message from a connected socket.
.TP
.BR recvfrom (3)
receive a message from a socket.
.TP
.BR recvmsg (3)
receive a message from a socket.
.TP
.BR send (3)
send a message on a socket.
.TP
.BR sendmsg (3)
send a message on a socket using a message structure.
.TP
.BR sendto (3)
sned a message on a socket.
.TP
.BR setsockopt (3)
set the socket options.
.TP
.BR shutdown (3)
shut down socket send and receive operations.
.TP
.BR socket (3)
create an endpoint for communications.
.TP
.BR socketpair (3)
create a pair of connected sockets.
.PD
.PP
See the associated manual page in section 3 of the manual for descriptions of
the socket API functions.
.\"
.\"
.SH IMPLEMENTATION
.PP
There are three traditional approaches to providing sockets over
.I STREAMS
as follows:
.\"
.\"
.SS SVR 3.2 Approach
The
.I SVR 3.2
.[
svr3
.]
approach consists of
a user-space library that performs most of the conversion between
.BR npi (7)
or
.BR tpi (7)
Streams and the sockets API using a cooperating
.BR sockmod (4)
.I STREAMS
module and the
.I SI_
input-output controls and other
.I STREAMS
facilities such as
.BR streamio (7),
.BR getmsg (2s)
and
.BR putmsg (2s).
.PP
Socket state information is maintained in user-space data structures
associated with each active file descriptor in a manner similar to the
.BR libxnet (3)
XTI/TLI library.
All socket API calls are implemented mostly in the user-space library with
some input-output control calls to
.BR sockmod (4).
.PP
This library approach uses the
.B /etc/netconfig
configuration file to map
.BR socket (3)
parameters onto
.I STREAMS
pseudo-devices within the library.
.PP
This approach is problematic for handling threaded programs and issues during
.BR fork (2).
.B @PACKAGE_TITLE@
implements compatibile
.I SI_
input-output controls in the
.BR sockmod (4)
module and the
.BR socksys (4)
driver, and provides a usable
.B /etc/netconfig
file, to support older binaries statically compiled against such a library;
but it does not implement such a library directly.
.PP
See
.BR sockmod (4)
for more information.
.\"
.\"
.SS SVR 4 Approach
The
.I SVR 4
.[
svr4
.]
approach consists of
a user-space library that performs few conversions between
.BR tpi (7)
Streams and the sockets API using a cooperating
.BR sockmod (4)
.I STREAMS
module and the
.I SIOCSOCKSYS
input-output control with
.I SO_
commands.
.PP
Socket state information is maintained in kernel data structures
associated with each Stream.
All socket API calls are implemented as emulated system calls using
.B SIOCSOCKSYS
input-output control calls to
.BR sockmod (4).
.PP
This library approach also uses the
.B /etc/netconfig
configuration file to map
.BR socket (3)
parameters onto
.I STREAMS
pseudo-devices within the library.
This approach does not support
.BR npi (7)
Streams.
.PP
This approach is less problematic from the point of view that state
information is maintained within the kernel instead of the library; however,
use of
.I SIOCSOCKSYS
input-output controls for each socket API call is not very efficient on older
kernels.
The
.BR sockmod (4)
module is still used and still supports the older
.I SI_
input-output controls for backward compatibility with the
.I SVR 3.2
.[
svr3
.]
approach above.
.PP
See
.BR sockmod (4)
for more information.
.\"
.\"
.SS \fISolaris\fP\(rg \fI6\fP Approach
The
.IR Solaris "\(rg " 6
.[
soldoc
.]
approach consists of
a user-space library that performs almost no conversions between
.BR tpi (7)
Streams and the sockets API using a special
.BR specfs (5)
filesystem and a special
.B so_create
system call beneath
.BR socket (3).
.PP
The
.B @PACKAGE_TITLE@
package emulates the
.B so_create
system call as an input-output control against the
.BR socksys (4)
driver.
.PP
Socket state information is maintained in kernel data structures
associated with each Stream.
All socket API calls (with the exception of
.BR socket (3))
are implemented as system calls.
.PP
This library approach uses the
.B /etc/sock2path
configuration file and the
.BR soconfig (8)
utility to create socket to stream mappings within the kernel.
This approach does not support
.BR npi (7)
Streams.
.PP
Aside from the method used to create a socket, this method is
indistinguishable from native
.B Linux
sockets and is the preferred approach.
.B @PACKAGE_TITLE@
uses this approach under
.B Linux Fast-STREAMS
.[
lfs
.]
which is capable of supporting it.
.PP
See
.BR socksys (4)
for more information.
.\"
.\"
.\".SH NOTES
.\".PP
.\"
.\"
.\".SH CAVEATS
.\".PP
.\"
.\"
.SH "RETURN VALUE"
.PP
Upon success, most sockets API library functions return zero (0) or a positive
integer reflecting a returned parameter value.
Upon failure,
sockets API library functions return -1 and set
.IR t_errno (3)
to an applicable error number.
.\"
.\"
.\".SH "ERROR HANDLING"
.\".PP
.\"
.\"
.SH ERRORS
.PP
The errors that can be returned by socket API library calls are detailed in
.IR errno (3).
.\"
.\"
.SH "SEE ALSO"
.PP
.BR accept (3),
.BR bind (3),
.BR connect (3),
.BR getpeername (3),
.BR getsockname (3),
.BR getsockopt (3),
.BR listen (3),
.BR recv (3),
.BR recvfrom (3),
.BR recvmsg (3),
.BR send (3),
.BR sendmsg (3),
.BR sendto (3),
.BR setsockopt (3),
.BR shutdown (3),
.BR socket (3),
.BR socketpair (3).
.\"
.\"
.SH BUGS
.PP
The sockets API library has been tested with the
.BR test-sock (8)
and
.BR test-sock_thr (8)
conformance test suites.
.PP
The sockets API library has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
The sockets API library functions are compatible with
.IR "SVR 4.2" ,
.[
svr42
.]
.IR "XNS 5.2" ,
.[
xns98
.]
.IR SUSv2 ,
.[
susv2
.]
.IR POSIX/SUSv3 .
.[
susv3
.]
.PP
Compatibility is tested using the
.BR test-sock (8)
and
.BR test-sock_thr (8)
test case executables and the
.I @PACKAGE_TITLE@
.B autotest
test suite.
.\"
.\"
.SH CONFORMANCE
.PP
.IR "The Open Group XNS Issue 5.2" ,
.[
xns98
.]
.IR "Single UNIX\(rg Specification Version 2" ,
.[
susv2
.]
.IR "Single UNIX\(rg Specification Version 3" .
.[
susv3
.]
.PP
Conformance is tested using the
.BR test-sock (8)
and
.BR test-sock_thr (8)
test case executables and the
.I @PACKAGE_TITLE@
.B autotest
test suite.
.\"
.\"
.SH HISTORY
.PP
The Sockets API interface first appeared in
.BR SVR4 .
.[
svr4
.]
.\"
.\"
.SH VERSIONS
.PP
This manpage was written for the OpenSS7 Sockets Library.
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


iperf/src/sockets.c
/*****************************************************************************

 @(#) $RCSfile$ $Name$($Revision$) $Date$

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2007  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date$ by $Author$

 -----------------------------------------------------------------------------

 $Log$
 *****************************************************************************/

#ident "@(#) $RCSfile$ $Name$($Revision$) $Date$"

static char const ident[] = "$RCSfile$ $Name$($Revision$) $Date$";

/*--------------------------------------------------------------- 
 * Copyright (c) 1999,2000,2001,2002,2003                              
 * The Board of Trustees of the University of Illinois            
 * All Rights Reserved.                                           
 *--------------------------------------------------------------- 
 * Permission is hereby granted, free of charge, to any person    
 * obtaining a copy of this software (Iperf) and associated       
 * documentation files (the "Software"), to deal in the Software  
 * without restriction, including without limitation the          
 * rights to use, copy, modify, merge, publish, distribute,        
 * sublicense, and/or sell copies of the Software, and to permit     
 * persons to whom the Software is furnished to do
 * so, subject to the following conditions: 
 *
 *     
 * Redistributions of source code must retain the above 
 * copyright notice, this list of conditions and 
 * the following disclaimers. 
 *
 *     
 * Redistributions in binary form must reproduce the above 
 * copyright notice, this list of conditions and the following 
 * disclaimers in the documentation and/or other materials 
 * provided with the distribution. 
 * 
 *     
 * Neither the names of the University of Illinois, NCSA, 
 * nor the names of its contributors may be used to endorse 
 * or promote products derived from this Software without
 * specific prior written permission. 
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES 
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
 * NONINFRINGEMENT. IN NO EVENT SHALL THE CONTIBUTORS OR COPYRIGHT 
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, 
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 
 * ________________________________________________________________
 * National Laboratory for Applied Network Research 
 * National Center for Supercomputing Applications 
 * University of Illinois at Urbana-Champaign 
 * http://www.ncsa.uiuc.edu
 * ________________________________________________________________ 
 *
 * socket.c
 * by Mark Gates <mgates@nlanr.net>
 * -------------------------------------------------------------------
 * set/getsockopt and read/write wrappers
 * ------------------------------------------------------------------- */

#include "headers.h"
#include "util.h"

#ifdef __cplusplus
extern "C" {
#endif

/* -------------------------------------------------------------------
 * If inMSS > 0, set the TCP maximum segment size  for inSock.
 * Otherwise leave it as the system default.
 * ------------------------------------------------------------------- */

const char warn_mss_fail[] = "\
WARNING: attempt to set TCP maxmimum segment size to %d failed.\n\
Setting the MSS may not be implemented on this OS.\n";

const char warn_mss_notset[] =
"WARNING: attempt to set TCP maximum segment size to %d, but got %d\n";

void setsock_tcp_mss( int inSock, int inMSS ) {
#ifdef TCP_MAXSEG
    int rc;
    int newMSS;
    Socklen_t len;

    assert( inSock != INVALID_SOCKET );

    if ( inMSS > 0 ) {
        /* set */
        newMSS = inMSS;
        len = sizeof( newMSS );
        rc = setsockopt( inSock, IPPROTO_TCP, TCP_MAXSEG, (char*) &newMSS,  len );
        if ( rc == SOCKET_ERROR ) {
            fprintf( stderr, warn_mss_fail, newMSS );
            return;
        }

        /* verify results */
        rc = getsockopt( inSock, IPPROTO_TCP, TCP_MAXSEG, (char*) &newMSS, &len );
        WARN_errno( rc == SOCKET_ERROR, "getsockopt TCP_MAXSEG" );
        if ( newMSS != inMSS ) {
            fprintf( stderr, warn_mss_notset, inMSS, newMSS );
        }
    }
#endif
} /* end setsock_tcp_mss */

/* -------------------------------------------------------------------
 * returns the TCP maximum segment size
 * ------------------------------------------------------------------- */

int getsock_tcp_mss( int inSock ) {
    int theMSS = 0;

#ifdef TCP_MAXSEG
    int rc;
    Socklen_t len;
    assert( inSock >= 0 );

    /* query for MSS */
    len = sizeof( theMSS );
    rc = getsockopt( inSock, IPPROTO_TCP, TCP_MAXSEG, (char*) &theMSS, &len );
    WARN_errno( rc == SOCKET_ERROR, "getsockopt TCP_MAXSEG" );
#endif

    return theMSS;
} /* end getsock_tcp_mss */

/* -------------------------------------------------------------------
 * Attempts to reads n bytes from a socket.
 * Returns number actually read, or -1 on error.
 * If number read < inLen then we reached EOF.
 *
 * from Stevens, 1998, section 3.9
 * ------------------------------------------------------------------- */

ssize_t readn( int inSock, void *outBuf, size_t inLen ) {
    size_t  nleft;
    ssize_t nread;
    char *ptr;

    assert( inSock >= 0 );
    assert( outBuf != NULL );
    assert( inLen > 0 );

    ptr   = (char*) outBuf;
    nleft = inLen;

    while ( nleft > 0 ) {
        nread = read( inSock, ptr, nleft );
        if ( nread < 0 ) {
            if ( errno == EINTR )
                nread = 0;  /* interupted, call read again */
            else
                return -1;  /* error */
        } else if ( nread == 0 )
            break;        /* EOF */

        nleft -= nread;
        ptr   += nread;
    }

    return(inLen - nleft);
} /* end readn */

/* -------------------------------------------------------------------
 * Attempts to write  n bytes to a socket.
 * returns number actually written, or -1 on error.
 * number written is always inLen if there is not an error.
 *
 * from Stevens, 1998, section 3.9
 * ------------------------------------------------------------------- */

ssize_t writen( int inSock, const void *inBuf, size_t inLen ) {
    size_t  nleft;
    ssize_t nwritten;
    const char *ptr;

    assert( inSock >= 0 );
    assert( inBuf != NULL );
    assert( inLen > 0 );

    ptr   = (char*) inBuf;
    nleft = inLen;

    while ( nleft > 0 ) {
        nwritten = write( inSock, ptr, nleft );
        if ( nwritten <= 0 ) {
            if ( errno == EINTR )
                nwritten = 0; /* interupted, call write again */
            else
                return -1;    /* error */
        }

        nleft -= nwritten;
        ptr   += nwritten;
    }

    return inLen;
} /* end writen */

#ifdef __cplusplus
} /* end extern "C" */
#endif
